### PR TITLE
Prepare AngularXSS for Rails 7

### PR DESCRIPTION
**Pull request by makmic**
Some changes were required as Rails 7 defines separate methods für explicit and implicit string conversion:
* 6.1: https://github.com/rails/rails/blob/v6.1.4.4/activesupport/lib/active_support/core_ext/string/output_safety.rb
* 7.0: https://github.com/rails/rails/blob/v7.0.0/activesupport/lib/active_support/core_ext/string/output_safety.rb
The CHANGELOG however does not say anything about it.

### CODE REVIEW

File: lib/angular_xss/safe_buffer.rb [hunk 1]
```diff
@@ -1,20 +1,36 @@
...
 
-    alias_method :html_escape_interpolated_argument_without_angular_xss, :html_escape_interpolated_argument
-    alias_method :html_escape_interpolated_argument, :html_escape_interpolated_argument_with_angular_xss
+    if private_method_defined?(:html_escape_interpolated_argument)
+      alias_method :explicit_html_escape_interpolated_argument_without_angular_xss, :html_escape_interpolated_argument
+      alias_method :html_escape_interpolated_argument, :explicit_html_escape_interpolated_argument_with_angular_xss
+    elsif private_method_defined?(:explicit_html_escape_interpolated_argument)
 
-  end
+      def implicit_html_escape_interpolated_argument_with_angular_xss(arg)
```
Comment (foobear) [inline lib/angular_xss/safe_buffer.rb]: Using a separate `if` statement which checks for the "implicit" version might be clearer:

```ruby
    if private_method_defined?(:html_escape_interpolated_argument)
      alias_method :html_escape...
    elsif private_method_defined?(:explicit_html_escape_interpolated_argument)
      alias_method :explicit_html_escape...
    end

    if private_method_defined?(:implicit_html_escape_interpolated_argument)
      def implicit_html_escape_interpolated_argument_with_angular_xss(arg)
        ...
      end

      alias_method :implicit_html_escape...
    end
```

File: lib/angular_xss/safe_buffer.rb [hunk 2]
```diff
@@ -1,20 +1,36 @@
...
     private
 
-    def html_escape_interpolated_argument_with_angular_xss(arg)
+    def explicit_html_escape_interpolated_argument_with_angular_xss(arg)
       if arg.html_safe?
         arg
       else
-        html_escape_interpolated_argument_without_angular_xss(AngularXss::Escaper.escape(arg))
+        explicit_html_escape_interpolated_argument_without_angular_xss(AngularXss::Escaper.escape(arg))
       end
```
Comment (foobear) [inline lib/angular_xss/safe_buffer.rb]: The implementation on at least Rails 6.1 and 7.0 is:

```ruby
        (!html_safe? || arg.html_safe?) ? arg : CGI.escapeHTML(arg.to_s)
```

We're missing the `!html_safe?` check on `self` here which might actually break supported features (I'm thinking of concatting `nil` onto a SafeBuffer which works in vanilla Rails; have not tested if that is broken here).

Might be worth checking if we can safely live without it, or if we need to implement a version switch, e.g. "Rails < 6 = only `arg.html_safe?` vs Rails 6+ = `!html_safe? || arg.html_safe?`).
If we have to do that, we could maybe do everything via version checks instead of checking for defined methods.

Comment (makmic) [reply lib/angular_xss/safe_buffer.rb]: The following snippet was always part of ActiveSupport and was just renamed:
```
 (!html_safe? || arg.html_safe?) ? arg : CGI.escapeHTML(arg.to_s)
# name in rails 7:explicit_html_escape_interpolated_argument
# name in rails 6.1 up to 4.1: html_escape_interpolated_argument
```
Even Rails 4.1 had the exact same implementation: https://github.com/rails/rails/blob/4-1-stable/activesupport/lib/active_support/core_ext/string/output_safety.rb#L233

I could not produce an issue with the current code (without `!html_safe?`):
```ruby
# vanilla
require 'active_support'
irb(main):009:0> str = ActiveSupport::SafeBuffer.new("")
=> ""
irb(main):010:0> str << nil
=> ""
irb(main):011:0> str << "<html>"
=> "&lt;html&gt;"

## with the monkey patch
require './lib/angular_xss/safe_buffer.rb'
require 'lib/angular_xss/escaper.rb'
str2 = ActiveSupport::SafeBuffer.new("")
=> ""
irb(main):015:0> str2 << nil
=> ""
str2 = ActiveSupport::SafeBuffer.new("")
=> ""
irb(main):015:0> str2 << nil
=> ""
 str << "{{foo}}"
=> "{{ $root.DOUBLE_LEFT_CURLY_BRACE }}foo}}"
```

..but I don't see why we should keep it this way either. I'll replace both usages with `(!html_safe? || arg.html_safe?)` and trust green specs on this one

**Review by makmic — COMMENTED** [review]
Applied three changes:
* replaced the `arg.html_safe?` check with `!html_safe? || arg.html_safe?`
* added an separate definition check/code block for implicit_*
* reworked the code to interpolate definition names based on a base variable `html_escape = :html_escape_interpolated_argument` - I found it too hard to parse without it

**Review by foobear — APPROVED** [review]
Code looks good to me!
The boilerplate is a bit larger than expected, and I have no good solution on how to reduce it, so feel free to merge. :)
